static/js/message_fetch.js

```diff
@@ -334,13 +334,13 @@ exports.start_backfilling_messages = function () {
 exports.initialize = function () {
     // get the initial message list
     function load_more(data) {
-        // If we received the initially selected message, select it on the client side,
-        // but not if the user has already selected another one during load.
-        //
-        // We fall back to the closest selected id, as the user may have removed
-        // a stream from the home before already
+        // If we haven't selected a message in the home view yet, and
+        // the home view isn't empty, we select the anchor message here.
         if (home_msg_list.selected_id() === -1 && !home_msg_list.empty()) {
-            home_msg_list.select_id(page_params.pointer,
+            // We fall back to the closest selected id, as the user
+            // may have removed a stream from the home view while we
+            // were loading data.
+            home_msg_list.select_id(data.anchor,
```

Comment:
This seems to be the change that causes the node test to first start failing, and when you debug the test, you can see that `data.anchor` is clearly `undefined` when we call `home_msg.select_id`, which eventually causes this line to fail:

~~~ js
156         function convert_id(str_id) {
157             const id = parseFloat(str_id);
158             if (isNaN(id)) {
159                 blueslip.fatal("Bad message id " + str_id);
160             }
161             return id;
162         }
~~~

So to be on the paranoid side, Tim, you should ask yourself if there's ever reason `data.anchor` could be `undefined` here during normal operation.

But I think it's a test artifact.  Still investigating...

Reply:
The first substantive line of the failing test is this:

~~~ js
const response = initialize_data.initial_fetch.resp;
~~~

And here's the relevant setup data:

~~~ js
139     initial_fetch: {
140         req: {
141             anchor: 444,
142             num_before: 200,
143             num_after: 200,
144             client_gravatar: true,
145         },
146         resp: {
147             messages: message_range(201, 801),
148             found_newest: false,
149         },
150     },
~~~

I am 60% sure that when I wrote the original tests here, we weren't yet echoing `anchor` from the server.  And I'm 99.99% sure we weren't using it the `message_fetch` code, so it would have been just kind of noise there...

Reply:
But with your change this simple tweak should get the test passing.  I'm not gonna push it, since it's easy to apply, and it is worth a quick conversation about whether we can always rely on `anchor` being there in server responses.  

~~~
$ git diff
diff --git a/frontend_tests/node_tests/message_fetch.js b/frontend_tests/node_tests/message_fetch.js
index 1c55d18..4a06b6d 100644
--- a/frontend_tests/node_tests/message_fetch.js
+++ b/frontend_tests/node_tests/message_fetch.js
@@ -146,6 +146,7 @@ const initialize_data = {
         resp: {
             messages: message_range(201, 801),
             found_newest: false,
+            anchor: 444,
         },
     },
~~~

Reply:
I actually think it's possibly a little confusing that when we send in `anchor`, it can now have all of these special values, but in the response, I think we're relying on the invariant that it's always an integer.  I'm gonna try to look at the host code a bit--I wasn't really in the loop for the latest round of changes to anchor/oldest/newest/10000000000000000/etc.

Reply:
I guess it was indeed me who first had us echo `anchor` in the server response during March 2018 in a4a8527ec5de32a91bcda45eefdd9e4770d381ec to fix #8639

Reply:
I put a `console.log(data.anchor)` statement inside of the success callback for `message_fetch.load_messages`, and I always get a number when clicking around, but I'm not sure I'm hitting every edge case.  I do get `10000000000000000` as a value sometimes, which could possibly be problematic?

Reply:
I'm not sure the webapp actually uses any of the magic string values for anchor yet, which would explain why anchor always comes back as a number.  Still digging...

Reply:
Oh, and the server does convert anchor to integer values with the following code at the very top of the view, so we can rely on integers:

~~~
anchor = parse_anchor_value(anchor_val, use_first_unread_anchor_val)
~~~

The only problem is that we still might receive this value back:

~~~
 772     if anchor_val == "newest":
 773         return LARGER_THAN_MAX_MESSAGE_ID
~~~

Which might be fine.

Reply:
Yeah, I think the API should, both by design and implementation, always return an integer for the anchor that was actually used.  These strings are just a request for the server to figure out the anchor to use for us because it has better access to data.